### PR TITLE
Support custom components in React "Link"s

### PR DESCRIPTION
**Pull request by gwn**
Current behavior is to render an HTML anchor element but it's a
common requirement to customize it. For instance if one wants to
implement a button link they need to fall back to the imperative
use of the router:

    <Button onClick={() => router.navigate('foo')} />

While it could be done in (arguably) a much nicer way if links
supported custom components:

    <Link routeName='foo' component='button' />

This is a common React idiom.

### CODE REVIEW

File: packages/react-router5/modules/BaseLink.ts [hunk 1]
```diff
@@ -1,22 +1,23 @@
...
-    onMouseOver?: MouseEventHandler<HTMLAnchorElement>
+    onClick?: MouseEventHandler<HTMLElement>
+    onMouseOver?: MouseEventHandler<HTMLElement>
     successCallback?(state?: State): void
     errorCallback?(error?: any): void
     target?: string
     route?: State
     previousRoute?: State
     router: Router
+    component?: any
```
Comment (gwn) [inline packages/react-router5/modules/BaseLink.ts]: I don't speak Typescript so I just left "any" here. This should have the type of "a string OR a function OR an instance of Component". I'm asking for help on this one.

**Comment by gwn** [comment]
I also don't have deep knowledge about the workings of the `types/BaseLink.d.ts ` file. I did replace all occurrences of `HTMLAnchorElement` with `HTMLElement` and hope this does it.

**Comment by troch** [comment]
Thanks for your PR. I'm not in favour of allowing this kind of logic, because it will make the implementation of `BaseLink` more complex. It is not very difficult to create your own link component using available hooks (or HoCs), and that's actually my recommendation: make your own components!

**Comment by gwn** [comment]
All right, thanks for your time and feedback.

**Pull request closed without merging**